1-clients/src/main/scala/com/snowplowanalytics/piinguin/client/PiinguinClient.scala

```diff
@@ -0,0 +1,247 @@
+    val responseObserver = getResponseObserverChangeRecord(responseQueue)
+    val requestObserver  = stub.createPiiRecords(responseObserver)
+    kvPairs.dequeue
+      .observeAsync(maxQueued)(Sink(requestSinkCreatePiiRecord(requestObserver)))
+      .compile
+      .drain
+  }
+
+  /**
+   * EXPERIMENTAL
+   * Given two fs2 queues, it consumes requests from one and enqueues responses in the other.
+   *
+   * @param maxQueued the maximum requests to be queued before the request queue blocks
+   * @param keys request queue
+   * @param responseQueue response queue
+   */
+  def deleteRecordsStreamingIO(maxQueued: Int,
+                               keys: Queue[IO, String],
+                               responseQueue: Queue[IO, Either[String, String]]): IO[Unit] = {
+    val responseObserver = getResponseObserverChangeRecord(responseQueue)
+    val requestObserver  = stub.deletePiiRecords(responseObserver)
+    keys.dequeue
+      .observeAsync(maxQueued)(Sink(requestSinkDeletePiiRecord(requestObserver)))
+      .compile
+      .drain
+  }
+
+  /**
+   * EXPERIMENTAL
+   * Given two fs2 queues, it consumes requests from one and enqueues responses in the other.
+   *
+   * @param maxQueued the maximum requests to be queued before the request queue blocks
+   * @param keysAndJustification request queue
+   * @param responseQueue response queue
+   */
+  def readRecordsStreamingIO(maxQueued: Int,
+                             keysAndJustification: Queue[IO, (String, LawfulBasisForProcessing)],
+                             responseQueue: Queue[IO, Either[String, PiiRecord]]): IO[Unit] = {
+    val responseObserver = getResponseObserverReadRecord(responseQueue)
+    val requestObserver  = stub.readPiiRecords(responseObserver)
+    keysAndJustification.dequeue
+      .observeAsync(maxQueued)(Sink(requestSinkReadPiiRecord(requestObserver)))
+      .compile
+      .drain
+  }
+
+  private def getResponseObserverChangeRecord(responseQueueIO: Queue[IO, Either[String, String]]) =
+    new StreamObserver[ChangeRecordResponse] {
+
+      def enqueue(either: Either[String, String]): Unit =
+        async.unsafeRunAsync(responseQueueIO.enqueue1(either))(_ => IO.unit)
+
+      override def onError(t: Throwable): Unit = enqueue(err(s"${t.getMessage}"))
+
+      override def onCompleted(): Unit = {}
+
+      override def onNext(value: ChangeRecordResponse): Unit =
+        enqueue({
+          value match {
+            case ChangeRecordResponse(None)        => err("Failed to get a response")
+            case ChangeRecordResponse(Some(error)) => if (error.isSuccess) ok(error.message) else err(error.message)
+          }
+        })
+    }
+  private def getResponseObserverReadRecord(responseQueueIO: Queue[IO, Either[String, PiiRecord]]) =
+    new StreamObserver[ReadPiiRecordResponse] {
+
+      def enqueue(either: Either[String, PiiRecord]): Unit =
+        async.unsafeRunAsync(responseQueueIO.enqueue1(either))(_ => IO.unit)
```

Comment:
I think we're losing the point of encapsulating effects here

Reply:
This is flowing two existing examples. As I said this is, for me, something that most likely be completely rewritten as we don't have enough knowledge and experience.